lib/semian/lru_hash.rb

```diff
@@ -0,0 +1,75 @@
+class LRUHash
+  extend Forwardable
+  def_delegators :@table, :size, :count, :empty?
+  attr_reader :table, :minimum_time_in_lru
+
+  class NoopMutex
+    def synchronize(*)
+      yield
+    end
+  end
+
+  [:keys, :clear].each do |attribute|
+    define_method :"#{attribute}" do
+      @lock.synchronize { @table.public_send(attribute) }
+    end
+  end
+
+  def initialize(minimum_time_in_lru:)
+    @minimum_time_in_lru = minimum_time_in_lru
+    @table = {}
+    @lock =
+      if Semian.thread_safe?
+        Mutex.new
+      else
+        NoopMutex.new
+      end
+  end
+
+  def set(key, resource)
+    delete(key)
+    @lock.synchronize do
+      @table[key] = resource
```

Comment:
Now that `delete` is thread safe, I'm trying to figure out if we still need to synchronize on this.